qa/tasks/cephfs/test_client_recovery.py

```diff
@@ -567,3 +567,33 @@ def test_abort_conn(self):
         self.assert_session_state(gid, "open")
         time.sleep(session_timeout * 1.5)  # Long enough for MDS to consider session stale
         self.assert_session_state(gid, "stale")
+
+    def test_dont_mark_unresponsive_client_stale(self):
+        """
+        Test that an unresponsive client holding no caps is evicted directly,
+        without being marked as stale even after session_timeout, at
+        session_autoclose.
+
+        To conduct this test we need more that two clients since a single
+        client is never evcited by MDS.
+        """
+        SESSION_TIMEOUT = 30
+        SESSION_AUTOCLOSE = 50
+        mount_a_gid = self.mount_a.get_global_id()
+        mount_a_pid = self.mount_a.get_client_pid()
+        self.fs.set_var('session_timeout', SESSION_TIMEOUT)
+        self.fs.set_var('session_autoclose', SESSION_AUTOCLOSE)
+        self.assert_session_count(2, self.fs.mds_asok(['session', 'ls']))
+
+        self.mount_a.run_shell(['mkdir', 'dir'])
+        self.mount_a.send_signal('sigstop')
+        time.sleep(SESSION_TIMEOUT + 2)
+        self.assert_session_state(mount_a_gid, "open")
+
+        # actual test part
+        self.mount_b.run_shell(['stat', 'dir'])
+        time.sleep(SESSION_AUTOCLOSE - SESSION_TIMEOUT + 2)
+        self.assert_session_count(1, self.fs.mds_asok(['session', 'ls']))
+
```

Comment:
`self.mount_a.send_signal('SIGCONT')` to confirm mount_a cleans up after itself.

Reply:
Do you mean write this at the end of the test? This crashes the test -

```
  File "/home/user/master/qa/tasks/cephfs/test_client_recovery.py", line 603, in test_dont_mark_unresponsive_client_stale
    self.mount_a.send_signal('sigcont')
  File "/home/user/master/qa/tasks/cephfs/mount.py", line 524, in send_signal
    self.run_shell(['kill', '-{0}'.format(signal), self.get_client_pid()])
  File "/home/user/master/qa/tasks/cephfs/fuse_mount.py", line 471, in get_client_pid
    status = self.admin_socket(['status'])
  File "/home/user/master/qa/tasks/cephfs/fuse_mount.py", line 446, in admin_socket
    stdout=StringIO(), timeout=(15*60))
  File "../qa/tasks/vstart_runner.py", line 341, in run
    proc.wait()
  File "../qa/tasks/vstart_runner.py", line 179, in wait
    raise CommandFailedError(self.args, self.exitstatus)
CommandFailedError: Command failed with status 22: ['./bin/ceph', '--admin-daemon', '/tmp/ceph-asok.zUi3YV//client.0.17646.asok', 'status']

2019-06-13 19:59:33,621.621 ERROR:__main__:Error in test 'test_dont_mark_unresponsive_client_stale (tasks.cephfs.test_client_recovery.TestClientRecovery)', going interactive
Ceph test interactive mode, use ctx to interact with the cluster, press control-D to exit...
>>> 
```

Reply:
`self.mount_b.run_shell('kill -sigcont ' + mount_a_pid)` and `self.mount_a.send_signal('sigcont')` are equivalent right? The former works but latter doesn't.

Reply:
Your PR adds `send_signal` :laughing:!

It looks like the reason you see different behavior is you're sending the signal to different mounts: `mount_a` vs `mount_b`.

Reply:
> Your PR adds `send_signal` laughing!

Damn, how did forget that? :P

> It looks like the reason you see different behavior is you're sending the signal to different mounts: `mount_a` vs `mount_b`.

Not _to_ different mounts but _via_ different mounts. I was sending `sigcont` to `mount_a` via `mount_a` which won't work. So I need to use `mount_b` to send `sigcont` to `mount_a`. Marking this conversation resolved since I've made this change in the code yesterday.

Reply:
> > It looks like the reason you see different behavior is you're sending the signal to different mounts: `mount_a` vs `mount_b`.
> 
> Not _to_ different mounts but _via_ different mounts. I was sending `sigcont` to `mount_a` via `mount_a` which won't work. So I need to use `mount_b` to send `sigcont` to `mount_a`. Marking this conversation resolved since I've made this change in the code yesterday.

Um, I think it should work to send a signal to mount_a. If not, that may be some kind of bug in the test or in teuthology/paramiko.

BTW, there's no guarantee for the test that mount_a and mount_b are on the same machine. So, you may be sending a signal to a non-existent process on mount_b.

Reply:
> > > It looks like the reason you see different behavior is you're sending the signal to different mounts: `mount_a` vs `mount_b`.
> > 
> > 
> > Not _to_ different mounts but _via_ different mounts. I was sending `sigcont` to `mount_a` via `mount_a` which won't work. So I need to use `mount_b` to send `sigcont` to `mount_a`. Marking this conversation resolved since I've made this change in the code yesterday.
> 
> Um, I think it should work to send a signal to mount_a. If not, that may be some kind of bug in the test or in teuthology/paramiko.

Ok. I have got/had no idea about this.

> 
> BTW, there's no guarantee for the test that mount_a and mount_b are on the same machine. So, you may be sending a signal to a non-existent process on mount_b.

True. I am used to the thinking in the lines of `vstart_runner.py`'s model of testing.

Since you would know better and so that this PR could progress I've replaced the last line by `self.mount_a.send_signal('sigcont')`.

Reply:
I'll run it thorugh QA and see what happens.